src/main/java/ch/njol/skript/SkriptAddon.java

```diff
@@ -21,135 +21,216 @@
+			if (entry == null) // This entry has already been loaded before
+				continue;
+
+			String name = entry.getName();
+			if (name.startsWith(basePackage) && name.endsWith(".class")) {
+				boolean load = subPackages.length == 0;
+
+				if (load) { // No subpackages provided
+					load = recursive || StringUtils.count(name, '/') <= initialDepth;
+				} else {
+					for (String subPackage : subPackages) {
+						if (
+							// We also need to account for subpackage depths when not doing a recursive search
+							(recursive || StringUtils.count(name, '/') <= initialDepth + StringUtils.count(subPackage, '/'))
+							&& name.startsWith(subPackage, basePackage.length())
+						) {
 							load = true;
 							break;
 						}
 					}
-
-					if (load)
-						classNames.add(e.getName().replace('/', '.').substring(0, e.getName().length() - ".class".length()));
 				}
+
+				if (load)
+					classNames.add(name.replace('/', '.').substring(0, name.length() - ".class".length()));
 			}
+		}
 
-			classNames.sort(String::compareToIgnoreCase);
+		classNames.sort(String::compareToIgnoreCase);
 
-			for (String c : classNames) {
-				try {
-					Class.forName(c, true, plugin.getClass().getClassLoader());
-				} catch (ClassNotFoundException ex) {
-					Skript.exception(ex, "Cannot load class " + c + " from " + this);
-				} catch (ExceptionInInitializerError err) {
-					Skript.exception(err.getCause(), this + "'s class " + c + " generated an exception while loading");
-				}
-			}
-		} finally {
+		for (String className : classNames) {
 			try {
-				jar.close();
-			} catch (IOException e) {}
+				Class<?> clazz = Class.forName(className, initialize, plugin.getClass().getClassLoader());
+				if (withClass != null)
+					withClass.accept(clazz);
+			} catch (ClassNotFoundException ex) {
+				Skript.exception(ex, "Cannot load class " + className);
+			} catch (ExceptionInInitializerError err) {
+				Skript.exception(err.getCause(), this + "'s class " + className + " generated an exception while loading");
+			}
 		}
+
 		return this;
 	}
+
+	/**
+	 * Loads all module classes found in the package search.
+	 * @param basePackage The base package to start searching in (e.g. 'ch.njol.skript').
+	 * @param subPackages Specific subpackages to search in (e.g. 'conditions').
+	 * If no subpackages are provided, all subpackages will be searched.
+	 * Note that the search will go no further than the first layer of subpackages.
+	 * @return This SkriptAddon.
+	 */
+	@SuppressWarnings("ThrowableNotThrown")
+	public SkriptAddon loadModules(String basePackage, String... subPackages) {
+		return loadClasses(basePackage, false, false, c -> {
+			if (Module.class.isAssignableFrom(c) && !c.isInterface() && !Modifier.isAbstract(c.getModifiers())) {
+				try {
+					((Module) c.getConstructor().newInstance()).register(this);
+				} catch (Exception e) {
+					Skript.exception(e, "Failed to load module " + c);
+				}
+			}
+		}, subPackages);
+	}
```

Comment:
I would prefer an alternative to this:
a) Code generation of module registration
b) Manual module registration

Reply:
If the addon discovery thing we talked about on Discord will become mainstream, a `modules: ["org.example.addon.math.MathModule", ...]` field can be added to `addon.json`.